File: PAT_Basic_Level_by_cpp/1037 在霍格沃茨找零钱/1037 在霍格沃茨找零钱.cpp
/*
 * 1037 在霍格沃茨找零钱
 *
 * 【实现原理】
 * 哈利波特世界中的货币体系：1 Galleon = 17 Sickle，1 Sickle = 29 Knut。
 * 给定应付价格（P）和实付金额（A），格式均为 G.S.K，
 * 计算找零（A - P），若实付不足则找零为负数。
 *
 * 【算法思路】
 * 1. 将应付金额和实付金额都转换为最小单位 Knut
 * 2. 计算差价 change = paid - price（单位为 Knut）
 * 3. 若 change < 0，先输出负号，再取绝对值
 * 4. 将 change 从 Knut 反向转换为 Galleon.Sickle.Knut 格式
 * 5. 按格式输出
 *
 * 【关键点】
 * - 统一转换为最小单位 Knut 再计算，避免借位问题
 * - 转换公式：totalKnut = (G * 17 + S) * 29 + K
 * - 反向转换：G = change / (17*29)，S = change % (17*29) / 29，K = change % 29
 *
 * 【复杂度分析】
 * - 时间复杂度：O(1)，仅常数次运算
 * - 空间复杂度：O(1)
 */

#include <iostream>
#include <cstdio>

using namespace std;

int main() {
    int pg, ps, pk, ag, as, ak;
    if (scanf("%d.%d.%d %d.%d.%d", &pg, &ps, &pk, &ag, &as, &ak) != 6) return 0;

    // 统一转换为最小单位 Knut
    int price = (pg * 17 + ps) * 29 + pk;  // 应付金额（Knut）
    int paid = (ag * 17 + as) * 29 + ak;   // 实付金额（Knut）
    int change = paid - price;             // 找零（Knut）

    // 若找零为负，先输出负号再取绝对值
    if (change < 0) { cout << '-'; change = -change; }

    // 反向转换为 Galleon.Sickle.Knut 格式
    int g = change / (17 * 29);            // Galleon：1 G = 17*29 Knut
    change %= (17 * 29);                   // 剩余 Knut
    int s = change / 29;                   // Sickle：1 S = 29 Knut
    int k = change % 29;                   // 剩余 Knut

    printf("%d.%d.%d\n", g, s, k);
    return 0;
}

File: PAT_Basic_Level_by_ruby/1037 在霍格沃茨找零钱/1037 在霍格沃茨找零钱.rb
# 1037 在霍格沃茨找零钱
#
# 题目描述：
#   如果你是哈利·波特迷，你会知道魔法世界有它自己的货币系统 —— 就如海格告诉哈利的：“十七个银西可(Sickle)兑一个加隆(Galleon)，二十九个纳特(Knut)兑一个西可，很容易。”现在，给定哈利应付的价钱 P 和他实付的钱 A，你的任务是写一个程序来计算他应该被找的零钱。
#
# 实现原理：
#   使用特殊的货币进制（17和29）；计算找零金额（应付-实付）；处理负数情况和进位借位
#

GALLEON_IN_KNUT = 17 * 29

# 将 G.S.K 转为 Knut 总数
# @param str [String]
# @return [Integer]
def to_knut(str)
  g, s, k = str.split('.').map(&:to_i)
  (g * 17 + s) * 29 + k
end

if __FILE__ == $PROGRAM_NAME
  paid, price = gets.split
  diff = to_knut(paid) - to_knut(price)
  negative = diff.negative?
  diff = diff.abs

  galleon = diff / GALLEON_IN_KNUT
  remainder = diff % GALLEON_IN_KNUT
  sickle = remainder / 29
  knut = remainder % 29
  puts "#{negative ? '-' : ''}#{galleon}.#{sickle}.#{knut}"
end
